lang/py/avro/io.py

```diff
@@ -112,84 +164,218 @@ def __init__(self, fail_msg, writers_schema=None, readers_schema=None):
-    'map': lambda s, d: (isinstance(d, dict) and all(isinstance(key, unicode) for key in d) and
-                         all(validate(s.values, value) for value in d.values())),
-    'union': lambda s, d: any(validate(branch, d) for branch in s.schemas),
-    'record': lambda s, d: (isinstance(d, dict) and
-                            all(validate(f.type, d.get(f.name)) for f in s.fields) and
-                            {f.name for f in s.fields}.issuperset(d.keys())),
+def validate(expected_schema, datum, raise_on_error=False):
+    """Return True if the provided datum is valid for the excpted schema
+
+    If raise_on_error is passed and True, then raise a validation error
+    with specific information about the error encountered in validation.
+
+    :param expected_schema: An avro schema type object representing the schema against
+                            which the datum will be validated.
+    :param datum: The datum to be validated, A python dictionary or some supported type
+    :param raise_on_error: True if a AvroTypeException should be raised immediately when a
+                           validation problem is encountered.
+    :raises: AvroTyeException if datum is invalid and raise_on_error is True
+    :returns: True if datum is valid for expected_schema, False if not.
+    """
+    # use a FIFO queue to process schema nodes breadth first.
+    nodes = deque()
+    nodes.append(ValidationNode(expected_schema, datum, getattr(expected_schema, "name", None)))
+
+    while nodes:
+        current_node = nodes.popleft()
+
+        # _validate_node returns the node for iteration if it is valid. Or it returns None
+        valid_node = _validate_node(current_node)
+
+        if valid_node is not None:
+            # if there are children of this node to append, do so.
+            for child_node in _iterate_node(valid_node):
+                nodes.append(child_node)
+        else:
+            # the current node was not valid.
+            if raise_on_error:
+                raise AvroTypeException(current_node.schema, current_node.datum)
+            else:
+                # preserve the prior validation behavior of returning false when there are problems.
+                return False
+
+    return True
+
+
+def _validate_node(node):
+    """Return the result of applying the appropriate validator function to the provided node"""
+    # breakpoint()
+    key = (node.schema.type, getattr(node.schema, 'logical_type', None))
+    return _VALIDATORS.get(key, _default_validator)(node)
+
+
+def _iterate_node(node):
+    for item in _ITERATORS.get(node.schema.type, _default_iterator)(node):
+        yield ValidationNode(*item)
+
+
+#############
+# Iteration #
+#############
+
+def _default_iterator(_):
+    """Immediately raise StopIteration.
+
+    This exists to prevent problems with iteration over unsupported container types.
+    """
+    for blank in []:
```

Comment:
Why not `raise StopIteration` explicitly?

Reply:
👍

Reply:
Oh, wait.  Now I remember.  Mypy gave me problems with that as it wasn't really generator function.  Now that I've removed type hinting I can probably go back to that.

Reply:
Well, don't kill type hinting. Eventually we'll need it anyway. I'm slightly annoyed that mypy doesn't special case StopIteration, but PEP 20, etc etc.

How about `yield from ()`?

Reply:
Unfortunately, that syntax is illegal in Python 2.7 so as long as supporting it is required, we can't use that.  I went back to the first implementation in the long run.

Reply:
Sigh. Yeah, OK, that's one for the mailing list, isn't it?